### PR TITLE
server: export granular volume bytes and iops metrics

### PR DESCRIPTION
**Pull request by rhtyd**
*Problem*: The VM metrics has aggregated volume bytes read/write and iops metrics but not on per volume basis.
*Root Cause*: The volume stats sub-system is not used to export the metrics, the support is not available for VMware.
*Solution*: Use the volume stats sub-system and DB table to export the metrics via the listVolumes and listVolumeMetrics API, and implement support for VMware and fix issue with network and disk metrics in the VM metrics view.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

Screenshot from 2019-04-04 16-19-23

listVolume response will now have:
```
      "diskioread": 631,
      "diskiowrite": 130,
      "diskkbsread": 15084,
      "diskkbswrite": 135,
```

listVolumeMetrics response will have:
```
      "diskiopstotal": 761,
      "diskioread": 631,
      "diskiowrite": 130,
      "diskkbsread": 15084,
      "diskkbswrite": 135,

```

### CODE REVIEW

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 1]
```diff
@@ -3496,6 +3496,7 @@ protected Answer execute(GetVmStatsCommand cmd) {
     }
 
     protected Answer execute(GetVmDiskStatsCommand cmd) {
+        // TODO: implement for Vmware
```
Comment (rhtyd) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: Use/extend perfcounter, and see https://www.vmware.com/support/developer/converter-sdk/conv61_apireference/virtual_disk_counters.html for disk read/write and iops

**Comment by anuragaw** [comment]
@blueorangutan package

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 2]
```diff
@@ -3496,7 +3495,109 @@ protected Answer execute(GetVmStatsCommand cmd) {
...
+
+            final ManagedObjectReference dcMor = hyperHost.getHyperHostDatacenter();
+            final DatacenterMO dcMo = new DatacenterMO(getServiceContext(), dcMor);
+
+            final HashMap<String, List<VmDiskStatsEntry>> vmStatsMap = new HashMap<>();
+            for (final String vmName : cmd.getVmNames()) {
+                final VirtualMachineMO vmMo = dcMo.findVm(vmName);
+                final List<VmDiskStatsEntry> diskStats = new ArrayList<>();
+                for (VirtualDisk disk : vmMo.getAllDiskDevice()) {
+                    // TODO Get UUID, PATH of VirtualDisk
```
Comment (anuragaw) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: WIP - need to figure this out before testing end to end. CC/ @rhtyd

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware65u2

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-65u2

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware65u2

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-65u2

**Comment by rhtyd** [comment]
Fixed conflicts
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-65u2

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-65u2

**Pull request merged**